File: app/src/main/java/com/ginkage/ejlookup/SuggestionProvider.java
package com.ginkage.ejlookup;

import android.app.SearchManager;
import android.content.Context;
import android.content.SearchRecentSuggestionsProvider;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.text.TextUtils;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.TreeMap;
import java.util.TreeSet;

import static android.app.SearchManager.SUGGEST_COLUMN_QUERY;
import static android.app.SearchManager.SUGGEST_COLUMN_TEXT_1;
import static android.provider.BaseColumns._ID;

public class SuggestionProvider extends SearchRecentSuggestionsProvider {
    public static final String AUTHORITY = "com.ginkage.ejlookup.suggest";
    private static final String[] COLUMNS =
            new String[] {_ID, SUGGEST_COLUMN_TEXT_1, SUGGEST_COLUMN_QUERY};

    private final UriMatcher uriMatcher;
    private static final int URI_MATCH_SUGGEST = 1;

    public SuggestionProvider() {
        setupSuggestions(AUTHORITY, DATABASE_MODE_QUERIES);
        uriMatcher = new UriMatcher(UriMatcher.NO_MATCH);
        uriMatcher.addURI(AUTHORITY, SearchManager.SUGGEST_URI_PATH_QUERY, URI_MATCH_SUGGEST);
    }

    @Override
    public Cursor query(
            Uri uri,
            String[] projection,
            String selection,
            String[] selectionArgs,
            String sortOrder) {
        if (uriMatcher.match(uri) == URI_MATCH_SUGGEST) {
            String text = selectionArgs[0];
            if (!TextUtils.isEmpty(text)) {
                MatrixCursor result = new MatrixCursor(COLUMNS);
                ArrayList<String> suggest = getLookupResults(getContext(), text, text);
                if (suggest != null) {
                    int id = 0;
                    for (String res : suggest) {
                        id++;
                        result.addRow(new Object[] {id, res, res});
                    }
                    if (id > 0) {
                        return result;
                    }
                }
            }
        }
        return super.query(uri, projection, selection, selectionArgs, sortOrder);
    }

    private static int Tokenize(
            char[] text,
            int len,
            DictionaryFile fileIdx,
            HashMap<String, Integer> suggest,
            String task,
            long sugPos)
            throws IOException {
        int p, last = -1;

        for (p = 0; p < len; p++)
            if (Nihongo.letter(text[p])
                    || (text[p] == '\''
                            && p > 0
                            && p + 1 < len
                            && Nihongo.letter(text[p - 1])
                            && Nihongo.letter(text[p + 1]))) {
                if (last < 0) {
                    last = p;
                }
            } else if (last >= 0) {
                last = -1;
            }

        if (last >= 0) // Only search for the last word entered
        Traverse(new String(text, last, p - last), fileIdx, 0, "", suggest, task, sugPos);

        return last;
    }

    static class Pair implements Comparable<Pair> {
        final String line;
        final int freq;

        public final int compareTo(Pair other) {
            if (this.freq != other.freq) return other.freq - this.freq;
            return this.line.compareToIgnoreCase(other.line);
        }

        Pair(String line, int freq) {
            this.line = line;
            this.freq = freq;
        }
    }

    public static ArrayList<String> getLookupResults(Context context, String request, String task) {
        ArrayList<String> result = null;

        int maxsug =
                Integer.parseInt(
                        PreferenceManager.getDefaultSharedPreferences(context)
                                .getString(context.getString(R.string.setting_max_suggest), "10"));
        boolean romanize =
                EJLookupActivity.getPrefBoolean(
                        context.getString(R.string.setting_suggest_romaji), true);

        char[] text = new char[request.length()];
        request.getChars(0, request.length(), text, 0);

        String kanareq = Nihongo.Kanate(text);
        char[] kanatext = new char[kanareq.length()];
        kanareq.getChars(0, kanareq.length(), kanatext, 0);

        int qlen = Nihongo.Normalize(text);
        int klen = Nihongo.Normalize(kanatext);

        HashMap<String, Integer> suggest = new HashMap<>();

        int last = -1;
        try {
            long sugPos = 0;
            DictionaryFile fileIdx = new DictionaryFile(context.getAssets(), "suggest.dat");
            last = Tokenize(text, qlen, fileIdx, suggest, task, sugPos);
            if (!Arrays.equals(text, kanatext))
                Tokenize(kanatext, klen, fileIdx, suggest, task, sugPos);
            fileIdx.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }

        if (!suggest.isEmpty() && task.equals(EJLookupActivity.lastQuery)) {
            result = new ArrayList<>(10);
            TreeSet<Pair> freq = new TreeSet<>();
            for (String str : suggest.keySet()) {
                Integer n = suggest.get(str);
                freq.add(new Pair(str, n));
            }

            HashSet<String> duplicate = new HashSet<>();
            String begin = (last >= 0 ? request.substring(0, last) : "");
            for (Pair pit : freq)
                if (result.size() < maxsug) {
                    String str = pit.line;
                    String k = str;

                    if (romanize) {
                        int i;
                        boolean convert = true;
                        for (i = 0; i < str.length(); i++)
                            if (str.charAt(i) >= 0x3200) {
                                convert = false;
                                break;
                            }

                        if (convert) {
                            char[] txt = new char[str.length()];
                            str.getChars(0, str.length(), txt, 0);
                            k = Nihongo.Romanate(txt, 0, str.length() - 1);
                        }
                    }

                    if (!duplicate.contains(k)) {
                        result.add(begin + k);
                        duplicate.add(k);
                    }
                }
        }

        return result;
    }

    private static int betole(int p) {
        return ((p & 0x000000ff) << 24)
                + ((p & 0x0000ff00) << 8)
                + ((p & 0x00ff0000) >>> 8)
                + ((p & 0xff000000) >>> 24);
    }

    private static char shtoch(int p) {
        return (char) (((p & 0x000000ff) << 8) + ((p & 0x0000ff00) >>> 8));
    }

    private static boolean Traverse(
            String word,
            DictionaryFile fidx,
            long pos,
            String str,
            HashMap<String, Integer> suglist,
            String task,
            long sugPos)
            throws IOException {
        if (!task.equals(EJLookupActivity.lastQuery)) return false;
        fidx.seek(pos + sugPos);

        int tlen = fidx.readUnsignedByte();
        int c = fidx.readUnsignedByte();
        int freq = betole(fidx.readInt());
        boolean children = ((c & 1) != 0), unicode = ((c & 8) != 0), exact = !(word.equals(""));
        int match = 0, nlen = 0, wlen = word.length(), p;
        char ch;

        if (pos > 0) {
            String nword = "";

            if (tlen > 0) {
                if (unicode) {
                    char[] wbuf = new char[tlen];
                    for (c = 0; c < tlen; c++) wbuf[c] = shtoch(fidx.readUnsignedShort());
                    nword = new String(wbuf);
                } else {
                    byte[] wbuf = new byte[tlen];
                    fidx.read(wbuf, 0, tlen);
                    nword = new String(wbuf);
                }
            }

            nlen = nword.length();
            str += nword;

            if (exact) {
                word = word.substring(1);
                wlen--;

                while (match < wlen && match < nlen) {
                    if (word.charAt(match) != nword.charAt(match)) break;
                    match++;
                }
            }
        }

        if (match == nlen || match == wlen) {
            TreeMap<Integer, Character> cpos = new TreeMap<>();
            exact = exact && (match == nlen);

            if (children) // One way or the other, we'll need a full children list
            do { // Read it from this location once, save for later
                    ch = shtoch(fidx.readUnsignedShort());
                    p = betole(fidx.readInt());
                    if (match < wlen) { // (match == nlen), Traverse children
                        if (ch == word.charAt(match)) {
                            String newWord = word.substring(match); // Traverse children
                            return Traverse(
                                    newWord,
                                    fidx,
                                    (p & 0x7fffffff),
                                    str + ch,
                                    suglist,
                                    task,
                                    sugPos);
                        }
                    } else cpos.put(p & 0x7fffffff, ch);
                } while ((p & 0x80000000) == 0);

            if (match == wlen) {
                if (freq > 0 && !exact) {
                    Integer v = suglist.get(str);
                    if (v == null) v = 0;
                    v += freq;
                    suglist.put(str, v);
                }

                for (int child_pos :
                        cpos.keySet()) // Traverse everything that begins with this word
                Traverse("", fidx, child_pos, str + cpos.get(child_pos), suglist, task, sugPos);

                return true; // Got result
            }
        }

        return false; // Nothing found
    }
}

File: app/src/main/java/com/ginkage/ejlookup/SuggestionProvider.kt
package com.ginkage.ejlookup

import android.app.SearchManager
import android.content.Context
import android.content.SearchRecentSuggestionsProvider
import android.content.UriMatcher
import android.database.Cursor
import android.database.MatrixCursor
import android.net.Uri
import android.preference.PreferenceManager
import android.provider.BaseColumns
import android.text.TextUtils
import java.io.FileNotFoundException
import java.io.IOException
import java.io.UnsupportedEncodingException
import java.util.ArrayList
import java.util.Arrays
import java.util.HashMap
import java.util.HashSet
import java.util.TreeMap
import java.util.TreeSet

class SuggestionProvider : SearchRecentSuggestionsProvider() {
  private val uriMatcher: UriMatcher
  override fun query(
    uri: Uri,
    projection: Array<String>?,
    selection: String?,
    selectionArgs: Array<String>?,
    sortOrder: String?
  ): Cursor? {
    if (uriMatcher.match(uri) == URI_MATCH_SUGGEST) {
      val text = selectionArgs!![0]
      if (!TextUtils.isEmpty(text)) {
        val result = MatrixCursor(COLUMNS)
        val suggest = getLookupResults(context, text, text)
        if (suggest != null) {
          var id = 0
          for (res in suggest) {
            id++
            result.addRow(arrayOf<Any>(id, res, res))
          }
          if (id > 0) {
            return result
          }
        }
      }
    }
    return super.query(uri, projection, selection, selectionArgs, sortOrder)
  }

  internal class Pair(val line: String, val freq: Int) : Comparable<Pair> {
    override fun compareTo(other: Pair): Int {
      return if (freq != other.freq) other.freq - freq else line.compareTo(
        other.line,
        ignoreCase = true
      )
    }
  }

  companion object {
    const val AUTHORITY = "com.ginkage.ejlookup.suggest"
    private val COLUMNS = arrayOf(
      BaseColumns._ID,
      SearchManager.SUGGEST_COLUMN_TEXT_1,
      SearchManager.SUGGEST_COLUMN_QUERY
    )
    private const val URI_MATCH_SUGGEST = 1
    @Throws(IOException::class)
    private fun Tokenize(
      text: CharArray,
      len: Int,
      fileIdx: DictionaryFile,
      suggest: HashMap<String, Int>,
      task: String,
      sugPos: Long
    ): Int {
      var last = -1
      var p = 0
      while (p < len) {
        if (Nihongo.letter(text[p])
          || (text[p] == '\'' && p > 0 && p + 1 < len && Nihongo.letter(text[p - 1])
            && Nihongo.letter(text[p + 1]))
        ) {
          if (last < 0) {
            last = p
          }
        } else if (last >= 0) {
          last = -1
        }
        p++
      }
      if (last >= 0) // Only search for the last word entered
        Traverse(String(text, last, p - last), fileIdx, 0, "", suggest, task, sugPos)
      return last
    }

    fun getLookupResults(context: Context?, request: String, task: String): ArrayList<String>? {
      var result: ArrayList<String>? = null
      val maxsug =
        PreferenceManager.getDefaultSharedPreferences(context)
          .getString(context!!.getString(R.string.setting_max_suggest), "10")!!.toInt()
      val romanize: Boolean = EJLookupActivity.getPrefBoolean(
        context.getString(R.string.setting_suggest_romaji), true
      )
      val text = CharArray(request.length)
      request.toCharArray(text, 0, 0, request.length)
      val kanareq = Nihongo.Kanate(text)
      val kanatext = CharArray(kanareq.length)
      kanareq.toCharArray(kanatext, 0, 0, kanareq.length)
      val qlen = Nihongo.Normalize(text)
      val klen = Nihongo.Normalize(kanatext)
      val suggest = HashMap<String, Int>()
      var last = -1
      try {
        val sugPos: Long = 0
        val fileIdx = DictionaryFile(context.assets, "suggest.dat")
        last = Tokenize(text, qlen, fileIdx, suggest, task, sugPos)
        if (!Arrays.equals(text, kanatext)) Tokenize(kanatext, klen, fileIdx, suggest, task, sugPos)
        fileIdx.close()
      } catch (e: FileNotFoundException) {
        e.printStackTrace()
      } catch (e: UnsupportedEncodingException) {
        e.printStackTrace()
      } catch (e: IOException) {
        e.printStackTrace()
      }
      if (suggest.isNotEmpty() && task == EJLookupActivity.lastQuery) {
        result = ArrayList(10)
        val freq = TreeSet<Pair>()
        for (str in suggest.keys) {
          val n = suggest[str]
          freq.add(Pair(str, n!!))
        }
        val duplicate = HashSet<String?>()
        val begin = if (last >= 0) request.substring(0, last) else ""
        for (pit in freq) if (result.size < maxsug) {
          val str = pit.line
          var k: String? = str
          if (romanize) {
            var convert = true
            var i = 0
            while (i < str.length) {
              if (str[i].code >= 0x3200) {
                convert = false
                break
              }
              i++
            }
            if (convert) {
              val txt = CharArray(str.length)
              str.toCharArray(txt, 0, 0, str.length)
              k = Nihongo.Romanate(txt, 0, str.length - 1)
            }
          }
          if (!duplicate.contains(k)) {
            result.add(begin + k)
            duplicate.add(k)
          }
        }
      }
      return result
    }

    private fun betole(p: Int): Int {
      return ((p and 0x000000ff shl 24)
        + (p and 0x0000ff00 shl 8)
        + (p and 0x00ff0000 ushr 8)
        + (p and -0x1000000 ushr 24))
    }

    private fun shtoch(p: Int): Char {
      return ((p and 0x000000ff shl 8) + (p and 0x0000ff00 ushr 8)).toChar()
    }

    @Throws(IOException::class)
    private fun Traverse(
      what: String,
      fidx: DictionaryFile,
      pos: Long,
      sstr: String,
      suglist: HashMap<String, Int>,
      task: String,
      sugPos: Long
    ): Boolean {
      var word = what
      var str = sstr
      if (task != EJLookupActivity.lastQuery) return false
      fidx.seek(pos + sugPos)
      val tlen = fidx.readUnsignedByte()
      var c = fidx.readUnsignedByte()
      val freq = betole(fidx.readInt())
      val children = c and 1 != 0
      val unicode = c and 8 != 0
      var exact = word != ""
      var match = 0
      var nlen = 0
      var wlen = word.length
      var p: Int
      var ch: Char
      if (pos > 0) {
        var nword = ""
        if (tlen > 0) {
          if (unicode) {
            val wbuf = CharArray(tlen)
            c = 0
            while (c < tlen) {
              wbuf[c] = shtoch(fidx.readUnsignedShort())
              c++
            }
            nword = String(wbuf)
          } else {
            val wbuf = ByteArray(tlen)
            fidx.read(wbuf, 0, tlen)
            nword = String(wbuf)
          }
        }
        nlen = nword.length
        str += nword
        if (exact) {
          word = word.substring(1)
          wlen--
          while (match < wlen && match < nlen) {
            if (word[match] != nword[match]) break
            match++
          }
        }
      }
      if (match == nlen || match == wlen) {
        val cpos = TreeMap<Int, Char>()
        exact = exact && match == nlen
        if (children) // One way or the other, we'll need a full children list
          do { // Read it from this location once, save for later
            ch = shtoch(fidx.readUnsignedShort())
            p = betole(fidx.readInt())
            if (match < wlen) { // (match == nlen), Traverse children
              if (ch == word[match]) {
                val newWord = word.substring(match) // Traverse children
                return Traverse(
                  newWord,
                  fidx,
                  (p and 0x7fffffff).toLong(),
                  str + ch,
                  suglist,
                  task,
                  sugPos
                )
              }
            } else cpos[p and 0x7fffffff] = ch
          } while (p and -0x80000000 == 0)
        if (match == wlen) {
          if (freq > 0 && !exact) {
            var v = suglist[str]
            if (v == null) v = 0
            v += freq
            suglist[str] = v
          }
          for (child_pos in cpos.keys)  // Traverse everything that begins with this word
            Traverse("", fidx, child_pos.toLong(), str + cpos[child_pos], suglist, task, sugPos)
          return true // Got result
        }
      }
      return false // Nothing found
    }
  }

  init {
    setupSuggestions(AUTHORITY, DATABASE_MODE_QUERIES)
    uriMatcher = UriMatcher(UriMatcher.NO_MATCH)
    uriMatcher.addURI(AUTHORITY, SearchManager.SUGGEST_URI_PATH_QUERY, URI_MATCH_SUGGEST)
  }
}